src/react/helper/use-event-listener-ref.ts

```diff
@@ -0,0 +1,54 @@
+import { useCallback, useEffect, useRef } from 'react';
+
+/**
+ * The type of a callback function that can be stored in the reference.
+ */
+type Callback<CallbackArguments extends unknown[]> = (...args: CallbackArguments) => void;
+
+/**
+ * A hook that creates a reference to an event listener callback function. It is used to stabilize the reference
+ * across renders, so that listeners don't get unsubscribed and resubscribed on every render when passed in as a prop.
+ *
+ * For example, doing this:
+ *
+ * ```jsx
+ *   export function MySubscription() {
+ *     useHookWithListener(() => {})
+ *
+ *     return <div>My Subscription</div>
+ *   }
+ * ```
+ *
+ * Where the `useHookWithListener` hook is defined as:
+ *
+ * ```jsx
+ *  export function useHookWithListener(listener) {
+ *    const listenerRef = useEventListenerRef(listener);
+ *    useEffect(() => {
+ *      // Use the listenerRef
+ *    }, [listenerRef]);
+ * }
+ * ```
+ *
+ * Will ensure that the listener is not unsubscribed and resubscribed on every render (i.e. the useEffect will not be called
+ * on every render).
+ *
+ * We allow for the callback to be undefined, as callbacks in the majority of our hooks are optional.
+ *
+ * @internal
+ * @template Arguments - The type of arguments accepted by the callback function.
+ * @param callback - The callback function to be stored in the reference.
+ * @returns The callback function stored in the reference.
+ */
+export const useEventListenerRef = <Arguments extends unknown[]>(
+  callback?: Callback<Arguments>,
+): Callback<Arguments> | undefined => {
+  const ref = useRef<Callback<Arguments> | undefined>(callback);
```

Comment:
This never returns undefined now, but maybe it should if it gets called with undefined? To make sure we don't subscribe if not needed and unsubscribe if listener changes to `undefined`.

I guess adding a line
```
if (!callback) { return undefined; }
``` 
somewhere at the top? Not sure if before or after the `useRef` is ideal

Reply:
This was playing in my mind so I gave it a go...

We can't
```
if (bla) {
  useSomething()
}
```
but `useSomething` can return conditionally.

So all that means is that we have to move the condition to the bottom.
```
const stableCallback = useCallback(...) // as now but don't return
if (!callback) return undefined;
return stableCallback;
```

I've tried it here: https://github.com/ably/ably-chat-js/commit/c152f810f7c6e85c4eada3e015c9301e9d378aeb, seems to work

Reply:
Modified inline with this after standup discussion